### PR TITLE
Test support for plate export

### PR DESCRIPTION
**Pull request by labkey-chrisj**
#### Rationale
This change adds better support for creating plates and plate sets via java api, and it expands the utilities in `TestDataUtils `to include support for getting csv string from row maps, and for writing map data to tabular (csv, tsv) format

#### Related Pull Requests
tests in limsModules https://github.com/LabKey/limsModules/pull/278

#### Changes

- [x] add to `TestDataUtils`
- [x] Command,response,param for Plate creation
- [x] improve data parsing for command params

### CODE REVIEW

**Comment by labkey-chrisj** [comment]
oof, I've got to clean up my branch to get rid of all those merge-remote-tracking tracks

**Comment by labkey-tchad** [comment]
Preparing review...

**Comment by labkey-tchad** [comment]
## PR Review

<table>
<tr>
<tr><td> ⏱️&nbsp;<strong>Estimated&nbsp;effort&nbsp;to&nbsp;review [1-5]</strong></td><td>

4, due to the complexity and size of the PR, including multiple new classes and methods across different files. Understanding the context and ensuring the correctness of the implementation, especially around the JSON handling and API command/response structure, requires a thorough review.


</td></tr>
<tr><td> 🧪&nbsp;<strong>Relevant tests</strong></td><td>

No


</td></tr>
<tr><td rowspan=2> 🔍&nbsp;<strong>Possible issues</strong></td>
<td>

<strong>Possible Bug:</strong> In `PlateParams.java`, the method `toJSON` uses `_plateType` directly without checking its validity or ensuring it matches expected values. This could lead to incorrect JSON if `_plateType` is not one of the expected integers.</td></tr>
<tr>
<td>

<strong>Data Integrity:</strong> In `TestDataUtils.java`, when converting row maps to CSV or TSV strings, there's no handling for values that might contain the delimiter (comma for CSV, tab for TSV), which could result in malformed files.</td></tr>
<tr><td> 🔒&nbsp;<strong>Security concerns</strong></td><td>

No

</td></tr>
</table>


<details><summary> <strong>Code feedback:</strong></summary>

<hr><table><tr><td>relevant file</td><td>src/org/labkey/remoteapi/plate/PlateParams.java
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Consider validating `_plateType` before using it in `toJSON`. Ensure it matches one of the expected values (1-5) to prevent creating invalid JSON objects. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/LabKey/testAutomation/pull/1942/files#diff-39137bd4a201d4849538aa45f77bc6087ab812fbde5d2c686cf8aa5eacc41c4fR85'>json.put("plateType", _plateType);</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>src/org/labkey/test/util/TestDataUtils.java
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Implement escaping for values that contain the delimiter in `toTabular` method to ensure the integrity of generated CSV/TSV data. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/LabKey/testAutomation/pull/1942/files#diff-b0d90ee32c9f7d3fd019baebbbc87632d78a1b7b15066a2bb280d1bd53a447a6R78'>builder.append(String.join(delimiter, values));</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>src/org/labkey/remoteapi/plate/CreatePlateCommand.java
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Add input validation in `CreatePlateCommand` constructor to ensure `plateParams` is not null to avoid potential `NullPointerException` in `getJsonObject`. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/LabKey/testAutomation/pull/1942/files#diff-85ef7a346788148af9da0fa375552319fb155c049531bc8068f80d5ac7cc5e71R13'>_plateParams = plateParams;</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>src/org/labkey/remoteapi/plate/PlateParams.java
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

For better data encapsulation, consider changing public fields to private and provide public getter methods. For example, change `_assayType` to private and add a public getter for it. [medium]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/LabKey/testAutomation/pull/1942/files#diff-39137bd4a201d4849538aa45f77bc6087ab812fbde5d2c686cf8aa5eacc41c4fR14'>private String _assayType; // blank or Standard</a></td></tr></table><hr>

</details><hr>

<details> <summary><strong>✨ Review tool usage guide:</strong></summary><hr> 

**Overview:**
The `review` tool scans the PR code changes, and generates a PR review. The tool can be triggered [automatically](https://pr-agent-docs.codium.ai/usage-guide/automations_and_usage/#github-app-automatic-tools-when-a-new-pr-is-opened) every time a new PR is opened, or can be invoked manually by commenting on any PR.
When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L19) related to the review tool (`pr_reviewer` section), use the following template:
```
/review --pr_reviewer.some_config1=... --pr_reviewer.some_config2=...
```
With a [configuration file](https://pr-agent-docs.codium.ai/usage-guide/configuration_options/), use the following template:
```
[pr_reviewer]
some_config1=...
some_config2=...
```
    

<table><tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `review` tool can be configured with extra instructions, which can be used to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Specify the relevant sub-tool, and the relevant aspects of the PR that you want to emphasize.

Examples for extra instructions:
```
[pr_reviewer] # /review #
extra_instructions="""
In the 'possible issues' section, emphasize the following:
- Does the code logic cover relevant edge cases?
- Is the code logic clear and easy to understand?
- Is the code logic efficient?
...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.
        

</details></td></tr>

<tr><td><details> <summary><strong> How to enable\disable automation</strong></summary><hr>

- When you first install PR-Agent app, the [default mode](https://pr-agent-docs.codium.ai/usage-guide/automations_and_usage/#github-app-automatic-tools-when-a-new-pr-is-opened) for the `review` tool is:
```
pr_commands = ["/review", ...]
```
meaning the `review` tool will run automatically on every PR, with the default configuration.
Edit this field to enable/disable the tool, or to change the used configurations
        

</details></td></tr>

<tr><td><details> <summary><strong> Auto-labels</strong></summary><hr>

The `review` tool can auto-generate two specific types of labels for a PR:
- a `possible security issue` label, that detects possible [security issues](https://github.com/Codium-ai/pr-agent/blob/tr/user_description/pr_agent/settings/pr_reviewer_prompts.toml#L136) (`enable_review_labels_security` flag)
- a `Review effort [1-5]: x` label, where x is the estimated effort to review the PR (`enable_review_labels_effort` flag)


</details></td></tr>

<tr><td><details> <summary><strong> Extra sub-tools</strong></summary><hr>

The `review` tool provides a collection of possible feedbacks about a PR.
It is recommended to review the [possible options](https://pr-agent-docs.codium.ai/tools/review/#enabledisable-features), and choose the ones relevant for your use case.
Some of the feature that are disabled by default are quite useful, and should be considered for enabling. For example: 
`require_score_review`, `require_soc2_ticket`, `require_can_be_split_review`, and more.


</details></td></tr>

<tr><td><details> <summary><strong> Auto-approve PRs</strong></summary><hr>

By invoking:
```
/review auto_approve
```
The tool will automatically approve the PR, and add a comment with the approval.


To ensure safety, the auto-approval feature is disabled by default. To enable auto-approval, you need to actively set in a pre-defined configuration file the following:
```
[pr_reviewer]
enable_auto_approval = true
```
(this specific flag cannot be set with a command line argument, only in the configuration file, committed to the repository)


You can also enable auto-approval only if the PR meets certain requirements, such as that the `estimated_review_effort` is equal or below a certain threshold, by adjusting the flag:
```
[pr_reviewer]
maximal_review_effort = 5
```


</details></td></tr>



<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://pr-agent-docs.codium.ai/tools/) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 


</details></td></tr>

</table>

See the [review usage](https://pr-agent-docs.codium.ai/tools/review/) page for a comprehensive guide on using this tool.


</details>

File: src/org/labkey/remoteapi/plate/PlateParams.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.labkey.api.collections.RowMapFactory;
+import org.labkey.api.util.JsonUtil;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class PlateParams
```
Comment (labkey-tchad) [inline src/org/labkey/remoteapi/plate/PlateParams.java:5]: This class is trying to be two different things. Though they have some overlap, the creation form (`org.labkey.assay.PlateController.CreatePlateForm`) is distinct from the response you get back (`org.labkey.api.assay.plate.Plate`). Many values exist in one but not the other (e.g. "`data`", "`plateSetId`", "`plateId`") while "`plateType`" is a totally different data type.

There should be distinct classes on the test side as well.

File: src/org/labkey/test/util/TestDataUtils.java [hunk 2]
```diff
@@ -23,4 +24,60 @@ public static List<Map<String, Object>> rowMapsFromTsv(String tsvString) throws
...
+            List<String> values = new ArrayList<>();
+            for (String name : columns)
+            {
+                Object value = row.get(name);
+                values.add(value != null ? String.valueOf(value) : "");
+            }
+            builder.append(String.join(delimiter, values));
+            builder.append("\n");
+        }
+        return builder.toString();
```
Comment (labkey-tchad) [inline src/org/labkey/test/util/TestDataUtils.java:124]: Consider using the `TsvMapWriter` from the server API. I don't love referencing stuff in the server API but the dependency is already there so no harm in using it.
```suggestion
        StringBuilder builder = new StringBuilder();
        try (TSVWriter writer = new TSVMapWriter(rows))
        {
            writer.setHeaderRowVisible(includeHeaders);
            writer.setDelimiterCharacter(delimiter);
            writer.write(builder);
        }
        return builder.toString();
```
Check `org.labkey.api.data.TSVMapWriter.Tests` for more examples.

**Review by labkey-tchad — CHANGES_REQUESTED** [review]

File: src/org/labkey/remoteapi/plate/PlateParams.java (hunk 1 above)
Comment (labkey-chrisj) [reply src/org/labkey/remoteapi/plate/PlateParams.java:5]: Oh, just saw this.  Working on it now

Comment (labkey-tchad) [reply src/org/labkey/remoteapi/plate/PlateParams.java:5]: I suspect `PlateSetParams` has the same problem but I didn't catch it before.

File: src/org/labkey/test/util/TestDataUtils.java (hunk 2 above)
Comment (labkey-chrisj) [reply src/org/labkey/test/util/TestDataUtils.java:124]: When I try using `TsvMapWriter`, it builds fine but at runtime I am shown this error:
```

C:\dev\labkey\develop\server\testAutomation\src\org\labkey\test\util\TestDataUtils.java:114: error: cannot access HttpServletResponse
            writer.write(builder);
                  ^
  class file for jakarta.servlet.http.HttpServletResponse not found
``` 🙍

File: src/org/labkey/remoteapi/plate/PlateParams.java (hunk 1 above)
Comment (labkey-chrisj) [reply src/org/labkey/remoteapi/plate/PlateParams.java:5]: I implemented `PlateTypes `as an enum to use to constrain plate create inputs (valid plateType integers right now are 1-5), but when I look at  `org.labkey.api.assay.plate.Plate`'s return data, I wonder if it shouldn't be a bean to receive server response information.

I split out `CreatePlateParams `from `PlateParams`, now `CreatePlateCommand `takes a `CreatePlateCommand `and `PlateResponse `produces a `PlateParams`

File: src/org/labkey/test/util/TestDataUtils.java (hunk 2 above)
Comment (labkey-tchad) [reply src/org/labkey/test/util/TestDataUtils.java:124]: Ok. I guess there IS some harm in using the server api 😛

File: src/org/labkey/remoteapi/plate/PlateParams.java (hunk 1 above)
Comment (labkey-chrisj) [reply src/org/labkey/remoteapi/plate/PlateParams.java:5]: > I suspect `PlateSetParams` has the same problem but I didn't catch it before.

I just noticed `CreatePlateSetForm `will accept a List of `PlateManager.CreatePlateSetPlate`, that might be a much more streamlined way of creating plate sets with plates 🤔

**Comment by labkey-chrisj** [comment]
Oh, interesting. Intellij segregates changes you made from other peoples' changes when they're done in the context of its code-with-me feature. I had to commit them separately

File: src/org/labkey/test/util/TestDataUtils.java [hunk 3]
```diff
@@ -1,11 +1,16 @@
 package org.labkey.test.util;
 
 import org.apache.commons.io.IOUtils;
+import org.labkey.api.data.TSVMapWriter;
+import org.labkey.api.data.TSVWriter;
 import org.labkey.serverapi.reader.TabLoader;
+import org.openqa.selenium.devtools.v85.io.IO;
```
Comment (labkey-tchad) [inline src/org/labkey/test/util/TestDataUtils.java]: Looks like you need to clean up some imports

**Review by labkey-tchad — APPROVED** [review]
Looks like some imports are left from experimenting with `TsvWriter`

**Pull request merged**